### PR TITLE
rust: Migrate Ed25519 FFI to `cxx`

### PR DESCRIPTION
**Pull request by str4d**
Part of zcash/zcash#5716.

### CODE REVIEW

**Comment by str4d** [comment]
This will conflict with @nuttycom's work to deduplicate the transaction building logic in `z_mergetoaddress` and `z_shieldcoinbase`; I'm happy to wait on this PR until that work is merged.

**Comment by str4d** [comment]
[Rebased](https://github.com/zcash/zcash/compare/5f55b6913cf2114450a9f7785ed27c70223feb70..9ad6c7a975e4f3c575ce57e4b8d4287609c8957d) on current `master` to fix merge conflicts.

**Comment by str4d** [comment]
[Rebased on current `master`](https://github.com/zcash/zcash/compare/9ad6c7a975e4f3c575ce57e4b8d4287609c8957d..2d666fdad2d8ff6a46386db61b010ac6e7b34bd0) to fix merge conflicts.

**Comment by str4d** [comment]
[Rebased on `master`](https://github.com/zcash/zcash/compare/2d666fdad2d8ff6a46386db61b010ac6e7b34bd0..d6a5794cdcb5ee7caf896c0e2b31aa195c84dec5) now that #6377 has been merged.

**Comment by str4d** [comment]
[Rebased on `master`](https://github.com/zcash/zcash/compare/d6a5794cdcb5ee7caf896c0e2b31aa195c84dec5..b1dc94249c16756dcd51f6fd9a252152995f1624) to fix merge conflicts.

File: src/test/transaction_tests.cpp [hunk 1]
```diff
@@ -404,10 +404,10 @@ void test_simple_joinsplit_invalidity(uint32_t consensusBranchId, CMutableTransa
         CTransaction signTx(newTx);
         uint256 dataToBeSigned = SignatureHash(scriptCode, signTx, NOT_AN_INPUT, SIGHASH_ALL, 0, consensusBranchId, txdata);
 
-        assert(ed25519_sign(
-            &joinSplitPrivKey,
-            dataToBeSigned.begin(), 32,
-            &newTx.joinSplitSig));
+        ed25519::sign(
+            joinSplitPrivKey,
+            {dataToBeSigned.begin(), 32},
```
Comment (nuttycom) [inline src/test/transaction_tests.cpp:409]: It'd be clearer if this were using the `rust::Slice` constructor explicitly; this is a little subtle.

**Review by nuttycom — APPROVED** [review]
utACK

**Pull request merged**